Source/Scene/PointPrimitiveCollection.js

```diff
@@ -0,0 +1,976 @@
+
+        var pointPrimitives = this._pointPrimitives;
+        var pointPrimitivesLength = pointPrimitives.length;
+
+        updateMode(this, frameState);
+
+        pointPrimitives = this._pointPrimitives;
+        pointPrimitivesLength = pointPrimitives.length;
+        var pointPrimitivesToUpdate = this._pointPrimitivesToUpdate;
+        var pointPrimitivesToUpdateLength = this._pointPrimitivesToUpdateIndex;
+
+        var properties = this._propertiesChanged;
+
+        var createVertexArray = this._createVertexArray;
+
+        var vafWriters;
+        var pass = frameState.passes;
+        var picking = pass.pick;
+
+        // PERFORMANCE_IDEA: Round robin multiple buffers.
+        if (createVertexArray || (!picking && this.computeNewBuffersUsage())) {
+            this._createVertexArray = false;
+
+            for (var k = 0; k < NUMBER_OF_PROPERTIES; ++k) {
+                properties[k] = 0;
+            }
+
+            this._vaf = this._vaf && this._vaf.destroy();
+
+            if (pointPrimitivesLength > 0) {
+                // PERFORMANCE_IDEA:  Instead of creating a new one, resize like std::vector.
+                this._vaf = createVAF(context, pointPrimitivesLength, this._buffersUsage);
+                vafWriters = this._vaf.writers;
+
+                // Rewrite entire buffer if pointPrimitives were added or removed.
+                for (var i = 0; i < pointPrimitivesLength; ++i) {
+                    var pointPrimitive = this._pointPrimitives[i];
+                    pointPrimitive._dirty = false; // In case it needed an update.
+                    writePointPrimitive(this, context, vafWriters, pointPrimitive);
+                }
+
+                // Different pointPrimitive collections share the same index buffer.
+                this._vaf.commit();
+            }
+
+            this._pointPrimitivesToUpdateIndex = 0;
+        } else {
+            // PointPrimitives were modified, but none were added or removed.
+            if (pointPrimitivesToUpdateLength > 0) {
+                var writers = [];
+
+                if (properties[POSITION_INDEX] || properties[OUTLINE_WIDTH_INDEX] || properties[PIXEL_SIZE_INDEX]) {
+                    writers.push(writePositionSizeAndOutline);
+                }
+
+                if (properties[COLOR_INDEX] || properties[OUTLINE_COLOR_INDEX]) {
+                    writers.push(writeCompressedAttrib0);
+                }
+
+                if (properties[SHOW_INDEX] || properties[TRANSLUCENCY_BY_DISTANCE_INDEX]) {
+                    writers.push(writeCompressedAttrib1);
+                }
+
+                if (properties[SCALE_BY_DISTANCE_INDEX]) {
+                    writers.push(writeScaleByDistance);
+                }
+
+                vafWriters = this._vaf.writers;
+
+                if ((pointPrimitivesToUpdateLength / pointPrimitivesLength) > 0.1) {
+                    // If more than 10% of pointPrimitive change, rewrite the entire buffer.
```

Comment:
This is totally not in the scope of this PR, but if you have the time, it would be awesome to run some experiments to test this (would need to test different browsers and different GPUs) to see what a good cross-over point is.  We've had this heuristic for 3.5 years...